scripts/install_third_party.py

```diff
@@ -338,10 +339,74 @@ def download_manifest_files(filepath):
                     dependency_tar_root_name, dependency_target_root_name)
 
 
+def install_redis_cli():
+    """This installs the redis-cli to the local oppia third_party directory so
+    that development servers and backend tests can make use of a local redis
+    cache. Redis-cli installed here (redis-cli-6.0.6) is different from the
+    redis package installed in manifest.json (redis-3.5.3). The redis-3.5.3
+    package detailed in manifest.json is the Python library that allows users to
+    communicate with any Redis cache using Python. The redis-cli-6.0.6 package
+    installed in this function contains C++ scripts for the redis-cli and
+    redis-server programs detailed below.
+
+    The redis-cli program is the command line interface that serves up an
+    interpreter that allows users to connect to a redis database cache and
+    query the cache using the Redis CLI API. It also contains functionality to
+    shutdown the redis server. We need to install redis-cli separately from the
+    default installation of backend libraries since it is a system program and
+    we need to build the program files after the library is untarred.
+
+    The redis-server starts a Redis database on the local machine that can be
+    queried using either the Python redis library or the redis-cli interpreter.
+    """
+    try:
+        # Pipe output to /dev/null for silence in console.
+        null = python_utils.open_file('/dev/null', 'w')
+        subprocess.call([
+            './third_party/redis-cli-6.0.6/src/redis-cli', '--version'])
+        null.close()
+        python_utils.PRINT('Redis-cli is already installed.')
+    except OSError:
+        # The redis-cli is not installed, run the script to install it.
+        # NOTE: We do the installation here since we need to use make.
+        python_utils.PRINT('Installing redis-cli...')
+
+        download_and_untar_files(
+            'https://download.redis.io/releases/redis-6.0.6.tar.gz',
+            TARGET_DOWNLOAD_DIRS['backend'],
+            'redis-6.0.6', 'redis-cli-6.0.6')
+
+        # Temporarily change the working directory to redis-cli-6.0.6 so we can
+        # build the source code.
+        with common.CD('third_party/redis-cli-6.0.6/'):
+            # Build the scripts necessary to start the redis server.
+            # The make command only builds the C++ files in the src/ folder
+            # without modifying anything outside of the oppia root directory.
+            # It will build the redis-cli and redis-server files so that we can
+            # run the server from inside the oppia folder by executing the
+            # script src/redis-cli and src/redis-server.
+            subprocess.call(['make'])
```

Comment:
Should we do "make test"?

Reply:
I have this problem and I'm not sure how to solve it where the directory changes in Python are a little weird. If I run make from python it works but make test or any of the other commands(make distclean) etc. errors out with 'OSError: [Errno 2] No such file or directory' the reason how the test is executed.
@(cd ..; ./runtest)
But executing cd in Python doesn't seem to change the working directory so we can't run the test. make test works fine in regular terminal.
Not sure how I can fix this since there is some setup before that that needs to run before we start the Unix executable(or else I would've just run the executable from Python without calling 'make test'

Overall, I don't think it's necessary to run make test since it's just building C++ files in the src directory, not actually touching /etc/ or /usr/bin so the installation has no reason to fail(if it does fail to compile it should error out in make).

Reply:
See if you can temporarily update the make script to print the directory before and after the cd command, to see where it thinks it is.

Reply:
Yep I tried it and it thinks it's exactly where we changed the directory in the context switch, hence the actual file is not found.

Reply:
Sorry -- could you please be more specific? What is (a) the dir before the cd command, and (b) the dir after the cd command? I don't understand your last comment.

Reply:
Make test is added back in we reverted redis 6.0.6 to 6.0.5.